### ISSUE TITLE
Question: `Selectable` return value

### ISSUE THREAD

**Issue by boesing** [doctrine/collections#291]
Hey there,

as of now, I do only know about `Selectable#matching` returns `Collection` (which also includes `Selectable`).
So when we use this in our projects, we almost always do also assert that the returned value implements both `Collection&Selectable` (which is actually true for `ArrayCollection` in ORM context tho).

So I wondered if there are plans to either annotate something like:

```
interface Selectable
{
    /**
     * Selects all elements from a selectable that match the expression and
     * returns a new collection containing these elements.
     *
     * @return Collection<mixed>
     * @psalm-return Collection<TKey,T>&Selectable<TKey,T>
     */
    public function matching(Criteria $criteria);
}
```

or if the `Selectable` interface also extends the `Collection` interface.

----

So why do I ask for this?

We do have some use-cases where we do pre-filter collections form within a method.
So code like this is something which is "normal" in our codebase:

```
final class Foo
{
    /** @param Collection<array-key,Entity>&Selectable<array-key,Entity> */
    public function __construct(Collection $foo)
    {
        $this->foo = $foo;
    }

    /** @return Collection<array-key,Entity>&Selectable<array-key,Entity> */
    public function getValidEntries(DateTimeImmutable $from): Collection
    {
        $criteria = Criteria::create();
        $expression = Criteria::expr();
        return $this->foo->matching($expression->gte('validFrom', $from));
    }
   
}
```

Afterwards we do want to apply additional criterias (or not) so we need another `Selectable` (which is actually not returned by `Selectable#matching`).

Is there any best practice here?

**Comment by AndrolGenhald** [comment]
I run into this often enough with Psalm that I've stopped suppressing it (which risks suppressing an actual problem) and started using this utility function to wrap the result:

```php
/**
 * Workaround for https://github.com/doctrine/collections/issues/291
 *
 * @template TKey of array-key
 * @template TValue
 * @template TCollection of Collection<TKey, TValue>
 * @param TCollection $arg
 *
 * @return TCollection&Selectable<TKey, TValue>
 */
function assertSelectable($arg)
{
    if (!$arg instanceof Selectable) {
        throw new \InvalidArgumentException("Argument must be Selectable");
    }

    return $arg;
}
```

It would be nice if this could be fixed here somehow. I'd rather not just stub `Selectable` to change the return type since it's possible to return a `Collection` that doesn't implement `Selectable`.

**Comment by greg0ire** [comment]
I think it would make sense to annotate `matching` as described above given that all existing implementations indeed seem to fulfill that requirement. Please send a PR.

**Comment by AndrolGenhald** [comment]
Wouldn't that be a breaking change since users could implement `Selectable` and return a non-`Selectable` collection? Probably not very likely, but it's still a breaking change. Would that mean it has to wait for a major version, or would it go out with a new minor version?

**Comment by greg0ire** [comment]
Since this is only about static analysis, I'd say it's fine to pretend this was always what was intended, and ship it in a minor version.

**Comment by boesing** [comment]
Closed with #314
